osutil/disks/disks.go

```diff
@@ -336,59 +337,83 @@ func (d *disk) FindMatchingPartitionUUID(label string) (string, error) {
-		for {
-			currentMinor++
-			partMajMin := fmt.Sprintf("%d:%d", d.major, currentMinor)
-			props, err := udevProperties(filepath.Join("/dev/block", partMajMin))
-			if err != nil && strings.Contains(err.Error(), "Unknown device") {
-				// the device doesn't exist, we hit the end of the disk
-				break
-			} else if err != nil {
-				// some other error trying to get udev properties, we should fail
-				return "", fmt.Errorf("cannot get udev properties for partition %s: %v", partMajMin, err)
-			}
 
-			if props["DEVTYPE"] != "partition" {
-				// we ran into another disk, break out
-				break
-			}
+		// step 1. find the devpath for the disk, then glob for matching
+		//         devices using the devname in that sysfs directory
+		// step 2. iterate over all those devices and save all the ones that are
+		//         partitions using the partition sysfs file
+		// step 3. for all partition devices found, query udev to get the fs
+		//         label and partition uuid
+
+		udevProps, err := udevProperties(filepath.Join("/dev/block", d.Dev()))
+		if err != nil {
+			return "", err
+		}
+
+		// get the base device name
+		devName := udevProps["DEVNAME"]
+		if devName == "" {
+			return "", fmt.Errorf("cannot get udev properties for device %s, missing udev property \"DEVNAME\"", d.Dev())
+		}
+		// the DEVNAME as returned by udev includes the /dev/mmcblk0 path, we
+		// just want mmcblk0 for example
+		devName = filepath.Base(devName)
+
+		// get the device path in sysfs
+		devPath := udevProps["DEVPATH"]
+		if devPath == "" {
+			return "", fmt.Errorf("cannot get udev properties for device %s, missing udev property \"DEVPATH\"", d.Dev())
+		}
+
+		// glob for /sys/${devPath}/${devName}*
+		paths, err := filepath.Glob(filepath.Join(dirs.SysfsDir, devPath, devName+"*"))
+		if err != nil {
+			return "", fmt.Errorf("internal error getting udev properties for device %s: %v", err, d.Dev())
+		}
+
+		// Glob does not sort, so sort manually to have consistent tests
+		sort.Strings(paths)
 
-			// TODO: maybe save ID_PART_ENTRY_NAME here too, which is the name
-			//       of the partition. this may be useful if this function gets
-			//       used in the gadget update code
-			fsLabelEnc := props["ID_FS_LABEL_ENC"]
-			if fsLabelEnc == "" {
-				// this partition does not have a filesystem, and thus doesn't
-				// have a filesystem label - this is not fatal, i.e. the
-				// bios-boot partition does not have a filesystem label but it
-				// is the first structure and so we should just skip it
+		for _, path := range paths {
+			// check if this device is a partition
+			partFileBytes, err := ioutil.ReadFile(filepath.Join(path, "partition"))
+			if err != nil {
 				continue
 			}
-
-			partuuid := props["ID_PART_ENTRY_UUID"]
-			if partuuid == "" {
-				return "", fmt.Errorf("cannot get udev properties for partition %s, missing udev property \"ID_PART_ENTRY_UUID\"", partMajMin)
+			if strings.TrimSpace(string(partFileBytes)) == "1" {
```

Comment:
(nitpick) you could `if strings.TrimSpace(string(partFileBytes)) != "1" { continue }` to avoid a level of nesting here.

Reply:
yeah actually this bit is wrong because the `partition` files isn't just `0` or `1`, but I will send a patch fixing the usage of `partition` today

Reply:
this issue was fixed now